sql/catalyst/src/main/scala/org/apache/spark/sql/exceptions/SqlScriptingException.scala

```diff
@@ -0,0 +1,47 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.sql.exceptions
+
+import scala.jdk.CollectionConverters.MapHasAsJava
+
+import org.apache.spark.{SparkThrowable, SparkThrowableHelper}
+import org.apache.spark.sql.catalyst.trees.Origin
+import org.apache.spark.sql.exceptions.SqlScriptingException.errorMessageWithLineNumber
+
+class SqlScriptingException (
+    errorClass: String,
+    cause: Throwable,
+    origin: Option[Origin],
```

Comment:
If it's never None, shall we change the type to `origin: Origin`?

Reply:
If it needs to be optional, then we don't need it as all `TreeNode` implementations will capture the current Origin.

Reply:
It will be None when a future scripting exception does not want to show the line number, if it doesn't make sense in that context. If we removed it completely as you suggested, exceptions would not be able to choose whether or not to show the line number, unless we add a flag or something similar in the place of origin, however the current approach seems more flexible to me.

Reply:
> It will be None when a future scripting exception does not want to show the line number

Do you have an example? Let's make sure it's a real use case. We never hide the Origin in error messages if it's available.

Reply:
I don't have an example currently, @davidm-db suggested this so maybe he could provide more details.

Reply:
Don't have any example at the moment, we can make it non-optional for now and if we ever figure out there is a need, we will make it optional.

Reply:
@cloud-fan fixed, made it not an Option